From: Goldwyn Rodrigues <rgoldwyn@suse.com>
Subject: [PATCH] aio: mark AIO pseudo-fs noexec
References: bsc1023992, CVE-2016-10044
Patch-mainline: never, fixed differently by 22f6b4d

This is a different rendition of
22f6b4d ("aio: mark AIO pseudo-fs noexec")

Because of unavailability of super block internal only flags such as
SB_I_NOEXEC flags, we are using fs_flags to mark vfsmounts with MNT_NOEXEC.

Signed-off-by: Goldwyn Rodrigues <rgoldwyn@suse.com>

---
 fs/aio.c           |    1 +
 fs/namespace.c     |    3 +++
 include/linux/fs.h |    1 +
 3 files changed, 5 insertions(+)

--- a/fs/aio.c
+++ b/fs/aio.c
@@ -229,6 +229,7 @@ static int __init aio_setup(void)
 {
 	static struct file_system_type aio_fs = {
 		.name		= "aio",
+		.fs_flags	= FS_NOEXEC,
 		.mount		= aio_mount,
 		.kill_sb	= kill_anon_super,
 	};
--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -786,6 +786,9 @@ vfs_kern_mount(struct file_system_type *
 	if (flags & MS_KERNMOUNT)
 		mnt->mnt.mnt_flags = MNT_INTERNAL;
 
+	if (type->fs_flags & FS_NOEXEC)
+		mnt->mnt.mnt_flags |= MNT_NOEXEC;
+
 	root = mount_fs(type, flags, name, data);
 	if (IS_ERR(root)) {
 		free_vfsmnt(mnt);
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -1826,6 +1826,7 @@ struct file_system_type {
 #define FS_USERNS_MOUNT		8	/* Can be mounted by userns root */
 #define FS_USERNS_DEV_MOUNT	16 /* A userns mount does not imply MNT_NODEV */
 #define FS_RENAME_DOES_D_MOVE	32768	/* FS will handle d_move() during rename() internally. */
+#define FS_NOEXEC		131072 /* Should set NOEXEC flag */
 	struct dentry *(*mount) (struct file_system_type *, int,
 		       const char *, void *);
 	void (*kill_sb) (struct super_block *);
